### PR TITLE
fix(meeting-plugin): fix media ready event not triggering multiple times

### PR DESCRIPTION
**Pull request by arun3528**
Fixes #163778
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-163778
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-163247
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-163370

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 1]
```diff
@@ -1497,36 +1497,38 @@ export default class Meeting extends StatelessWebexPlugin {
...
+          EVENT_TRIGGERS.MEDIA_READY,
+          {
+            type: EVENT_TYPES.REMOTE_VIDEO,
+            stream: MediaUtil.createMediaStream([this.mediaProperties.remoteStream.getAudioTracks()[0], event.track])
+          }
+        );
+      }
+      else if (event.transceiver.mid === '2' && event.track) {
+        this.mediaProperties.setRemoteShare(MediaUtil.createMediaStream([event.track]));
         Trigger.trigger(
```
Comment (abailey-wwt) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1532]: If I understand correctly, now we are checking the media id of the RTCRtpTransceiver in order to fire the appropriate event. Also in the event of shareTrack we also additionally check event.track to make sure a MediaStreamTrack exists? If makes sense to me.

I just have one suggestion if you want. In case the way we trigger events changes, maybe we can not repeat that section but only include it once? In addition to using constants instead of magic strings.

Here's just a suggestion I had:

```javascript
  /**
   * Sets the remote stream on the class instance and emits and
   * event to developers
   * @param {Object} pc The remote stream peer connection
   * @returns {undefined}
   * @public
   * @memberof Meeting
   */
  setRemoteStream(pc) {
    if (!pc) {
      return;
    }
    // eslint-disable-next-line no-param-reassign
    pc.ontrack = (event) => {
      // eslint-disable-next-line no-warning-comments
      // TODO: It's possible for media to not be present
      // so we might need to either
      // A) wait until we have media flowing
      // B) trigger a second event when video is flowing
      LoggerProxy.logger.log(`Meeting:index#setRemoteStream --> ontrack event received for peerConnection: ${event}`);
      // const {transceiver} = event;
      const streams = event.streams[0];
      const MEDIA_ID = {
          AUDIO_TRACK: 0,
          VIDEO_TRACK: 1,
          SHARE_TRACK: 2
      }
      const trackMediaID = parseInt(event.transceiver.mid);
      let eventType = null;
      let eventStream = null;


      switch( trackMediaID ) {
          case MEDIA_ID.AUDIO_TRACK:
              eventType = EVENT_TYPES.REMOTE_AUDIO;
              eventStream = MediaUtil.createMediaStream([event.track]);
              this.mediaProperties.remoteStream = streams;
              break;
          case MEDIA_ID.VIDEO_TRACK:
              eventType = EVENT_TYPES.REMOTE_VIDEO;
              eventStream = MediaUtil.createMediaStream([this.mediaProperties.remoteStream.getAudioTracks()[0], event.track]);
              this.mediaProperties.setRemoteStream(MediaUtil.createMediaStream([this.mediaProperties.remoteStream.getAudioTracks()[0], event.track]));
              break;
          case MEDIA_ID.SHARE_TRACK:
              if ( event.track ) {
                eventType = EVENT_TYPES.REMOTE_SHARE;
                eventStream = this.mediaProperties.remoteShare;
                this.mediaProperties.setRemoteShare(MediaUtil.createMediaStream([event.track]));
              }
              break;
      }

      if (eventType && eventStream) {
        Trigger.trigger(
          this,
          {
            file: 'meeting/index',
            function: 'setRemoteStream:pc.ontrack'
          },
          EVENT_TRIGGERS.MEDIA_READY,
          {
            type: eventType,
            stream: eventStream
          }
        );
      }
    };
  }
```

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js [hunk 2]
```diff
@@ -22,6 +24,15 @@ MeetingsUtil.extractDestination = (destination, type) => {
   return dest;
 };
 
+// ** This is a workaround to fix Firefox and google android not downloading H264 codec on start
+MeetingsUtil.triggerH264Codec = async () => {
+  let testPeerconnection = new window.RTCPeerConnection();
+
+  await testPeerconnection.createOffer();
```
Comment (abailey-wwt) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:31]: I just had one thought here, would it make sense to test if the codec is available before creating a temporary  new peer connection and offer? I'm not sure of the penalties if any for doing so each time.

There's a `HTMLMediaElement.canPlayType()` that is supported by all browsers. Would it make sense to use it here? 

In case you want to look into it:
https://davidwalsh.name/detect-supported-video-formats-javascript
https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/canPlayType

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 1 above)
Comment (abailey-wwt) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1532]: The only other thing that jumped out at me was that you are pulling `this.mediaProperties.remoteStream.getAudioTracks()[0]` for the `EVENT_TYPES.REMOTE_VIDEO`. Why set the stream to the audio, if the event is remote video? Thanks.

File: .gitignore [hunk 3]
```diff
@@ -42,3 +42,6 @@ packages/node_modules/samples/index.html
 
 # patch-package
 patches.hash
+
+# Local Netlify folder
+.netlify
```
Comment (adamweeks) [inline .gitignore:47]: Please remove this change or add to a different commit.

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 4]
```diff
@@ -1497,36 +1497,38 @@ export default class Meeting extends StatelessWebexPlugin {
...
-        EVENT_TRIGGERS.MEDIA_READY,
-        {
-          type: EVENT_TYPES.REMOTE_VIDEO,
-          stream: MediaUtil.createMediaStream([this.mediaProperties.remoteStream.getVideoTracks()[0]])
-        }
-      );
-
-      if (this.mediaProperties.remoteStream.getVideoTracks()[1]) {
-        this.mediaProperties.setRemoteShare(MediaUtil.createMediaStream([this.mediaProperties.remoteStream.getVideoTracks()[1]]));
+      if (event.transceiver.mid === '0') {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1500]: Would like to see these strings replaced by a constant.

File: packages/node_modules/@webex/plugin-meetings/src/meetings/util.js [hunk 5]
```diff
@@ -22,6 +24,15 @@ MeetingsUtil.extractDestination = (destination, type) => {
   return dest;
 };
 
+// ** This is a workaround to fix Firefox and google android not downloading H264 codec on start
+MeetingsUtil.triggerH264Codec = async () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meetings/util.js:28]: This feature should be in its own commit or PR as it does not align with `fix(meeting-plugin): fix media ready event not triggering multiple times`

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 6]
```diff
@@ -1497,36 +1497,38 @@ export default class Meeting extends StatelessWebexPlugin {
       // const {transceiver} = event;
       const streams = event.streams[0];
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1498]: Can we rename this to `eventStream` instead? I was confused on line 1501 that did:
`this.mediaProperties.remoteStream = streams;`

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js (hunk 1 above)
Comment (arun3528) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1532]: fixing it

**Comment by arun3528** [comment]
@adamweeks  @abailey-wwt  have fixed the review comments asked for, Just changed the int media ID to string

File: packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js [hunk 7]
```diff
@@ -92,6 +93,10 @@ skipInBrowser(describe)('plugin-meetings', () => {
       assert.equal(StaticConfig.meetings.bandwidth.video, 500);
     });
 
+    it('Should call ', () => {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/test/unit/spec/meetings/index.js:96]:
```suggestion
    it('Should trigger h264 download', () => {
```

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 8]
```diff
@@ -1494,39 +1494,41 @@ export default class Meeting extends StatelessWebexPlugin {
...
+          break;
+        case MEDIA_ID.SHARE_TRACK:
+          if (event.track) {
+            eventType = EVENT_TYPES.REMOTE_SHARE;
+            eventStream = MediaUtil.createMediaStream([event.track]);
+            this.mediaProperties.setRemoteShare(eventStream);
+          }
+          break;
+        default: {
+          LoggerProxy.logger.log('Meeting:index#setRemoteStream no matching media track id');
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js:1527]:
```suggestion
          LoggerProxy.logger.log('Meeting:index#setRemoteStream --> no matching media track id');
```

**Comment by arun3528** [comment]
@adamweeks  fixed the review comments . Thank you
